### ISSUE TITLE
Support myst-parser

### ISSUE THREAD

**Issue by LecrisUT** [click-contrib/sphinx-click#127]
Currently the generated code is specific for `rst` documents such that the following does not work:
````markdown
```{click} module:parser
:prog: hello-world
```
````
There is a workaround using:
````markdown
```{eval-rst}
.. click:: [scikit_build_core.cli.main:skbuild](module:parser)
   :prog: hello-world
```
````
However it would be good to have built-in to use the appropriate parser (i.e. rst) for the generated code.

**Comment by stephenfin** [comment]
fwiw, I looked at this today. It seems `myst_parser` does not seem to be happy with our use of `sphinx.utils.nodes.nested_parse_with_titles`. That should return a load of docutils nodes but the output is off.

Here's an example using reStructuredText:

![Screen Shot 2024-04-06 at 10 28 29](https://github.com/click-contrib/sphinx-click/assets/1690835/5905db43-968f-493d-8c6e-e74c7e0241fe)

and here's the equivalent using Markdown:

![Screen Shot 2024-04-06 at 10 28 17](https://github.com/click-contrib/sphinx-click/assets/1690835/36af30b6-f7a7-4227-b284-95f96a485ba5)

You can see that all the directives are outputted in raw form rather than being processed and nested as expected. I'm not sure how you'd go about fixing this, tbh.

**Comment by stephenfin** [comment]
If anyone is interested in fixing this, there's a minimal reproducer [here](https://github.com/stephenfin/myst-sphinx-test).

**Comment by LecrisUT** [comment]
Yes, and I think that's expected, I've used something similar when messing around with autodoc extensions. For that case, it was fine because the nature of the docstring (rst or md fomat) is known by the user, amd they could fine-tune and select the appropriate parser with `eval-rst`.

The issue here are:
- the generated doc fragments do not account for which parser it is currently in, e.g.
  https://github.com/click-contrib/sphinx-click/blob/e8cf016c28920029f33597c43e02024b1fc61ba4/sphinx_click/ext.py#L344
  This could have an if statement to generate either format. It should be within an `with` to appropriately start-end the blocks
- the api is monolithic, but something along the lines of sphinx autodoc would probably be more manageable

I think the only relevant references here is `sphinx-autodoc2`, e.g. see their approach [here](https://github.com/sphinx-extensions2/sphinx-autodoc2/blob/19587f0f56a17f6f9cc1c2c5614ad8c98d7293ff/src/autodoc2/render/base.py#L274-L282)

**Comment by kdeldycke** [comment]
For the record, I just added [support of MyST syntax](https://kdeldycke.github.io/click-extra/sphinx.html#click-directives) for the `click:example` and `click:run` Sphinx directives.

It will be available in the upcoming [Click Extra 5.1.0](https://github.com/kdeldycke/click-extra/blob/main/changelog.md) (which is pending Click 8.2.2 release).
